clang/runtime/dpct-rt/include/dpct/sparse_utils.hpp

```diff
@@ -1458,6 +1520,243 @@ inline void csr2csc(sycl::queue queue, int m, int n, int nnz,
+                                      info_c->get_index_base(), row_ptr_c_buf,
+                                      col_ind_c_buf, val_c_buf);
+#else
+    oneapi::mkl::sparse::set_csr_data(
+        queue, *matrix_handle_a, rows_a, cols_a, info_a->get_index_base(),
+        (int *)row_ptr_a, (int *)col_ind_a, (Ty *)val_a);
+    oneapi::mkl::sparse::set_csr_data(
+        queue, *matrix_handle_b, rows_b, cols_b, info_b->get_index_base(),
+        (int *)row_ptr_b, (int *)col_ind_b, (Ty *)val_b);
+    oneapi::mkl::sparse::set_csr_data(queue, *matrix_handle_c, rows_c, cols_c,
+                                      info_c->get_index_base(), row_ptr_c,
+                                      (int *)nullptr, (Ty *)nullptr);
+#endif
+
+    matmat_desc = new oneapi::mkl::sparse::matmat_descr_t();
+    oneapi::mkl::sparse::init_matmat_descr(matmat_desc);
+    oneapi::mkl::sparse::set_matmat_data(
+        *matmat_desc, oneapi::mkl::sparse::matrix_view_descr::general, trans_a,
+        oneapi::mkl::sparse::matrix_view_descr::general, trans_b,
+        oneapi::mkl::sparse::matrix_view_descr::general);
+    desc->get_csrgemm_info_map().insert(std::make_pair(
+        args, descriptor::matmat_info(matrix_handle_a, matrix_handle_b,
+                                      nullptr, matrix_handle_c, matmat_desc)));
+  } else {
+    matrix_handle_a = Iter->second.matrix_handle_a;
+    matrix_handle_b = Iter->second.matrix_handle_b;
+    matrix_handle_c = Iter->second.matrix_handle_c;
+    matmat_desc = Iter->second.matmat_desc;
+  }
+
+#ifdef DPCT_USM_LEVEL_NONE
+#define __MATMAT(STEP, NNZ_C)                                                  \
+  oneapi::mkl::sparse::matmat(queue, *matrix_handle_a, *matrix_handle_b,       \
+                              *matrix_handle_c, STEP, *matmat_desc, NNZ_C,     \
+                              nullptr)
+#else
+#define __MATMAT(STEP, NNZ_C)                                                  \
+  oneapi::mkl::sparse::matmat(queue, *matrix_handle_a, *matrix_handle_b,       \
+                              *matrix_handle_c, STEP, *matmat_desc, NNZ_C,     \
+                              nullptr, {})
+#endif
+
+  __MATMAT(oneapi::mkl::sparse::matmat_request::work_estimation, nullptr);
+  queue.wait();
+
+  __MATMAT(oneapi::mkl::sparse::matmat_request::compute, nullptr);
+
+  int nnz_c_int = 0;
+#ifdef DPCT_USM_LEVEL_NONE
+  sycl::buffer<std::int64_t, 1> nnz_buf_c(1);
+  __MATMAT(oneapi::mkl::sparse::matmat_request::get_nnz, &nnz_buf_c);
+  nnz_c_int = nnz_buf_c.get_host_access(sycl::read_only)[0];
+#else
+  std::int64_t *nnz_c = sycl::malloc_host<std::int64_t>(1, queue);
+  __MATMAT(oneapi::mkl::sparse::matmat_request::get_nnz, nnz_c);
+  queue.wait();
+  nnz_c_int = *nnz_c;
+#endif
+
+  if (nnz_host_ptr) {
+    *nnz_host_ptr = nnz_c_int;
+  }
+  int last_element_of_row_ptr =
+      info_c->get_index_base() == oneapi::mkl::index_base::zero ? 0 : 1;
+  dpct::dpct_memcpy(row_ptr_c + m, &nnz_c_int, sizeof(int), host_to_device);
```

Comment:
I'm not sure what you are trying to do here.
1. `last_element_of_row_ptr` appears to be unused as the function ends just a couple of lines below.
1. In the `dpct_memcpy` call, it looks like you are trying to achieve `row_ptr_c[m] = nnz_c_int;`. Why is that needed? If a non-`nullptr` `row_ptr_c` was already provided in the API, `row_ptr_c[m]` is already `nnz_c_int + indexing` after the compute stage completes (current behavior of `matmat`).

Reply:
In fact, what you are doing here appears to be wrong for 1-based indexing since `row_ptr_c[m]` should be `nnz_c_int + indexing`.

Reply:
Yes, I am wrong here.
Actually, I want to do
```
  int last_element_of_row_ptr =
     nnz_c_int +  info_c->get_index_base() == oneapi::mkl::index_base::zero ? 0 : 1;
  dpct::dpct_memcpy(row_ptr_c + m, &last_element_of_row_ptr, sizeof(int), host_to_device);
```
here, but forgot to add the index_base in the current code.

Since `matmat` already do that, I can just remove those lines.